### PR TITLE
Fixes switch vibration

### LINKED ISSUE #27497: CupertinoSwitch haptic feedback triggered way too often
Introduced [here](https://github.com/flutter/flutter/commit/9f20bd6cb91bb61162ff60ba2ac84944f74b6e9b#diff-7fd54900390d62f0b4354d669adebcf4). The commit invokes haptic feedback when the onChanged *callback* is updated (which is every time widget hierarchy is rebuilt) instead of invoking haptic feedback when the actual value is changed.

@jslavitz 

```
[✓] Flutter (Channel master, v1.2.1-pre.67, on Mac OS X 10.14 18A389, locale en-US)
    • Flutter version 1.2.1-pre.67 at /Users/Matej/Projects/flutter/forked/flutter
    • Framework revision 42bbb441dd (3 months ago), 2018-11-21 16:40:41 +0100
    • Engine revision 15f2b92cce
    • Dart version 2.1.1 (build 2.1.1-dev.3.2 71bee8f05e)

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).
      If Android SDK has been installed to a custom location, set ANDROID_HOME to that location.
      You may also want to add it to your PATH environment variable.


[✓] iOS toolchain - develop for iOS devices (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • ios-deploy 1.9.4
    • CocoaPods version 1.5.3

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).

[!] IntelliJ IDEA Community Edition (version 2018.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.io/intellij-setup/#installing-the-plugins

[✓] VS Code (version 1.30.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 2.22.3

[✓] Connected device (1 available)
    • Matej’s iPhone X • f7e7ead1f9b5f3e2574cfbd9bc264fbcd5772a6c • ios • iOS 12.1.2
```

### PR DESCRIPTION
**Pull request by jslavitz**
Fixes #27497.

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/switch.dart [hunk 1]
```diff
@@ -263,6 +263,15 @@ class _RenderCupertinoSwitch extends RenderConstrainedBox {
       _positionController.forward();
     else
       _positionController.reverse();
+
+    switch(defaultTargetPlatform) {
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/switch.dart]: Worth double checking but I have a suspicion this isn't right. We should produce the knock when the knob hits the side, not when it starts moving. We should attach it to the _positionController's status handler.

File: packages/flutter/test/cupertino/switch_test.dart [hunk 2]
```diff
@@ -80,6 +80,78 @@ void main() {
...
+                ]
+              ),
+            );
+          },
+        ),
+      ),
+    );
+
+    await tester.tap(find.byKey(switchKey));
+    await tester.pumpAndSettle();
```
Comment (xster) [inline packages/flutter/test/cupertino/switch_test.dart]: In general, we want to avoid pumpAndSettle and be specific in the timing of the assertion. For instance, here, we're just checking that a haptic kick happened at some point. But we should more specifically test that the timing of the knock coincides with the timing of the knob reaching the edge (or whatever iOS does).

File: packages/flutter/test/cupertino/switch_test.dart [hunk 3]
```diff
@@ -80,6 +80,78 @@ void main() {
...
+
+    await tester.tap(find.byKey(switchKey2));
+    await tester.pumpAndSettle();
+
+    expect(log, hasLength(4));
+    expect(log[3], isMethodCall('HapticFeedback.vibrate', arguments: 'HapticFeedbackType.lightImpact'));
+    debugDefaultTargetPlatformOverride = null;
+  });
+
   testWidgets('Switch can drag (LTR)', (WidgetTester tester) async {
```
Comment (xster) [inline packages/flutter/test/cupertino/switch_test.dart]: Can you add another test checking that dragging the knob (instead of just tapping it) to the other edge also does whatever haptic feedback iOS does?

File: packages/flutter/lib/src/cupertino/switch.dart (hunk 1 above)
Comment (jslavitz) [reply packages/flutter/lib/src/cupertino/switch.dart]: Actually seems (unlike the picker) that it bumps immediately on tapUp/dragstart.

File: packages/flutter/test/cupertino/switch_test.dart [hunk 4]
```diff
@@ -80,6 +80,117 @@ void main() {
...
+    await tester.drag(find.byType(CupertinoSwitch), const Offset(30.0, 0.0));
+    expect(value, isTrue);
+    await tester.pump(const Duration(milliseconds: 100));
+
+    expect(log, hasLength(1));
+    expect(log[0], isMethodCall('HapticFeedback.vibrate', arguments: 'HapticFeedbackType.lightImpact'));
+    debugDefaultTargetPlatformOverride = null;
+  });
+
   testWidgets('Switch can drag (LTR)', (WidgetTester tester) async {
```
Comment (xster) [inline packages/flutter/test/cupertino/switch_test.dart]: Can you also check that the knock doesn't happen when the parent widget programmatically changes the current value. Only on user interaction?

File: packages/flutter/test/cupertino/switch_test.dart [hunk 5]
```diff
@@ -80,6 +80,117 @@ void main() {
...
+            );
+          },
+        ),
+      ),
+    );
+
+    await tester.tap(find.byKey(switchKey));
+    await tester.pump(const Duration(milliseconds: 100));
+
+    expect(log, hasLength(1));
```
Comment (xster) [inline packages/flutter/test/cupertino/switch_test.dart]: If we're saying that the tap happens right away, make this assertion before the 100ms pump and make sure there isn't another call after 100ms

File: packages/flutter/test/cupertino/switch_test.dart [hunk 6]
```diff
@@ -80,6 +80,117 @@ void main() {
...
+        ),
+      ),
+    );
+
+    await tester.drag(find.byType(CupertinoSwitch), const Offset(30.0, 0.0));
+    expect(value, isTrue);
+    await tester.pump(const Duration(milliseconds: 100));
+
+    expect(log, hasLength(1));
+    expect(log[0], isMethodCall('HapticFeedback.vibrate', arguments: 'HapticFeedbackType.lightImpact'));
```
Comment (xster) [inline packages/flutter/test/cupertino/switch_test.dart]: Same here. If the drag triggers it, we shouldn't need to wait 100ms before checking the knock happened.

File: packages/flutter/lib/src/cupertino/switch.dart [hunk 7]
```diff
@@ -390,8 +384,10 @@ class _RenderCupertinoSwitch extends RenderConstrainedBox {
   }
 
   void _handleDragStart(DragStartDetails details) {
-    if (isInteractive)
+    if (isInteractive) {
       _reactionController.forward();
+      _emitVibration();
```
Comment (xster) [inline packages/flutter/lib/src/cupertino/switch.dart]: Sorry this is becoming more complicated than initially thought. There's a bunch of nuances here it seems. On iOS, as far as I can tell, it's not possible to move enough to break the drag slop, but not enough to put the knob onto the other side and still get a knock. It may be possible here. Worth testing.

**Comment by xster** [comment]
LGTM

File: packages/flutter/lib/src/cupertino/switch.dart (hunk 7 above)
Comment (jslavitz) [reply packages/flutter/lib/src/cupertino/switch.dart]: I mean, the current implementation of the switch is wrong in respect to iOS 12. https://github.com/flutter/flutter/issues/27962 . Given that, I don't think there's really a completely 'correct' way to implement the vibration for the version of the switch that exists in flutter. No worries about the complication!
